### ISSUE TITLE
Different text depending on line break

### ISSUE THREAD

**Issue by wycats** [Marwes/pretty.rs#50]
I want to have a document that outputs a key and a value differently depending on whether the value first on one line:

```
description(string)
```

vs.

```
description
  table of
    row1 row2 row3 row4 
```

In other words, I'd like to change the document in the situation where a line break is deemed necessary. I'd be comfortable with many restrictions, and don't thing backtracking is necessary (iow what happens if the new document no longer needs a line break; too bad).

**Comment by Marwes** [comment]
I believe you want flat_alt https://github.com/Marwes/pretty.rs/blob/4a49547fc0a13357556af6fb94ea05459ff5d063/src/lib.rs#L595 ? Just pushed out 0.6 which contains that function (thought I had already released it)

**Comment by wycats** [comment]
Thanks!

**Comment by wycats** [comment]
Relatedly, I'm struggling a little bit with nest, which seems to require me to nest a document together, and let it possibly nest the second line and forward.

What I frequently want is: nest this whole document underneath this other document. Maybe it's obvious how this is supposed to work and I've just missed it?

**Comment by brendanzab** [comment]
elm-pretty seems to have [`hang`](https://package.elm-lang.org/packages/the-sett/elm-pretty-printer/latest/Pretty#hang), along with a bunch of other combinators. Not sure if that's what you are after though? I wonder if we could add some of them to pretty.rs? 🤔

**Comment by wycats** [comment]
The description of `hang` seems like what I'm experiencing with `nest`.

It seems like `nest` isn't exactly about nesting the document, but rather nesting the lines after the first line, but that frequently feels awkward to me.

**Comment by Marwes** [comment]
> What I frequently want is: nest this whole document underneath this other document. Maybe it's obvious how this is supposed to work and I've just missed it?

Might be that you are doing `doc1.append(Doc::space()).append(doc2.nest(..))` instead of `doc1.append(Doc::space().append(doc2).nest(..))` ?

**Comment by Marwes** [comment]
Adding `hang` would be possible but it requires the `Column` document primitive https://hackage.haskell.org/package/prettyprinter-1.5.1/docs/Data-Text-Prettyprint-Doc.html#v:column

**Comment by wycats** [comment]
@Marwes I'm pretty sure I'm not doing that, but let me whip up some examples that might help identify where my confusion is coming from.

**Comment by wycats** [comment]
@Marwes Here's an example: https://github.com/wycats/gitpod-examples/

The output looks like this:

![image](https://user-images.githubusercontent.com/4/69080600-26be5e80-09f1-11ea-9bd3-e5dd75eb5690.png)

What I'm trying to do is nest the child document, but as you can see it nests the second line of the child document. This is sort of fine, except that it requires me to know more about the per-line structure of the output than (1) I want to think about, and (2) pretty.rs wants me to think about

**Comment by Marwes** [comment]
Yeah, you need the column document. Started on it at https://github.com/Marwes/pretty.rs/pull/52 . It is going to be a bit awkward since the function in column needs to produce a new document and the layouting only works with references. So the owned document needs to be stored somewhere while the layouting is done.

**Comment by wycats** [comment]
For my use-case I'm not too worried about the ownership issues. That sounds perfect.

**Comment by Marwes** [comment]
The ownership issues are fortunately/unfortunately an implementation detail. You don't have to worry about it but the implementation does!

**Comment by Marwes** [comment]
Released in 0.7 . Let me know if it works for you!
